app/domain/authentication/authn/authenticator.rb

```diff
@@ -31,9 +31,13 @@ def valid?(input)
       end
 
       def credentials(input)
-        role_id = role_cls.roleid_from_username(input.account, input.username)
+        role_id = role_cache.roleid_from_username(input.account, input.username)
         credentials_cls[role_id]
       end
+
+      def role_cache
+        @role_class ||= role_cls.new
+      end
```

Comment:
Per our Slack convo, we can just inject `Role` here directly and do away with the cached version.  It turns out we can also fix my old `Dry::Struct` misuse here with very little effort.  Here's the updated class in its entirety.

Note I also I change the local variable names from `credentials` to `creds` so as to avoid any possible confusion with the `credentials` method.  

Clients of this class can still call it exactly as it was being called before.

Also, this may require minor modifications to a unit test or two somewhere.

```ruby
# frozen_string_literal: true

# default conjur authenticator
#
module Authentication
  module Authn

    class Authenticator

      def self.requires_env_arg?
        false
      end

      def initialize(role_cls: Role, credentials_cls: ::Credentials)
        @role_cls = role_cls
        @credentials_cls = credentials_cls
      end

      # Authenticates a Conjur using their username and password
      def login(input)
        creds = credentials(input)
        success = creds&.authenticate(input.password)

        success ? creds.api_key : nil
      end

      # Authenticates a Conjur role using its id and API key
      def valid?(input)
        creds = credentials(input)
        creds&.valid_api_key?(input.password)
      end

      def credentials(input)
        role_id = @role_cls.roleid_from_username(input.account, input.username)
        credentials_cls[role_id]
      end
    end
  end
end
```